editoast/src/views/pathfinding/catenaries.rs

```diff
@@ -0,0 +1,473 @@
+use crate::DbPool;
+use actix_web::{
+    dev::HttpServiceFactory,
+    get, services,
+    web::{block, Data, Json, Path},
+};
+use chashmap::CHashMap;
+use rangemap::RangeMap;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::ops::Range;
+
+pub fn routes() -> impl HttpServiceFactory {
+    services![catenaries_on_path,]
+}
+
+/// A struct used for the result of the catenaries_on_path endpoint
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+struct CatenaryRange {
+    begin: f64,
+    end: f64,
+    mode: String,
+}
+
+impl CatenaryRange {
+    fn list_from_range_map(range_map: &RangeMap<Float, String>) -> Vec<CatenaryRange> {
+        range_map
+            .iter()
+            .map(|(range, mode)| CatenaryRange {
+                begin: range.start.0,
+                end: range.end.0,
+                mode: mode.to_string(),
+            })
+            .collect()
+    }
+}
+
+/// A struct to make f64 Ord, to use in RangeMap
+#[derive(Debug, PartialEq, Copy, Clone, Serialize)]
+struct Float(f64);
+
+impl Float {
+    fn new(f: f64) -> Self {
+        assert!(f.is_finite());
+        Self(f)
+    }
+}
+
+impl From<f64> for Float {
+    fn from(f: f64) -> Self {
+        Self::new(f)
+    }
+}
+
+impl Ord for Float {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.partial_cmp(&other.0).unwrap()
+    }
+}
+
+impl PartialOrd for Float {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl Eq for Float {}
+
+/// Build a rangemap for each track section, giving the voltage for each range
+fn map_catenary_modes(
+    infra_cache: &InfraCache,
+    track_ids: HashSet<String>,
+) -> HashMap<String, RangeMap<Float, String>> {
+    let unique_catenary_ids = track_ids
+        .iter()
+        .flat_map(|track_id| {
+            infra_cache
+                .get_track_refs_type(track_id, ObjectType::Catenary)
+                .into_iter()
+        })
+        .map(|catenary_ref| &catenary_ref.obj_id)
+        .collect::<HashSet<_>>();
+
+    let mut res = HashMap::new();
+    for catenary_id in unique_catenary_ids {
+        let catenary = infra_cache
+            .catenaries()
+            .get(catenary_id)
+            .expect("catenary not found")
+            .unwrap_catenary();
+        for track_range in &catenary.track_ranges {
+            if track_ids.contains(&track_range.track as &str) {
+                let range = track_range.begin.into()..track_range.end.into();
+                let res_entry = res
+                    .entry(track_range.track.to_string())
+                    .or_insert_with(RangeMap::new);
+                if res_entry.overlaps(&range) {
+                    panic!(
+                        "Catenary {:?} overlaps with another catenary on range {:?}",
+                        catenary, track_range
+                    );
```

Comment:
We don't generate catenaries error yet. I don't know if overlapping catenary should be an error or warning. In all cases, an infra can have overlapping catenary and we should not panic.

We can return an error for now.